### PR TITLE
chore: Parallelize EnsureCleanState for e2e tests, adding timing information

### LINKED ISSUE #20967: Add more timing information about clean state preparation in e2e tests
# Summary

Right now the only reported timing info is for shell commands as well as total. The former is about 300-400ms, while the total can be 1.1-1.4sec. It would be great to get more details about various steps time to find targets to optimize.

Screenshot 2024-11-26 at 8 49 08 PM

# Motivation

Preparing a clean state takes 10-30% of each e2e test runtime, so finding optimization opportunities can significantly help with the runtime.

# Proposal

Use a code like
```
ts := stats.NewTimingStats()
...
ts.AddCheckpoint("some_step_ms")
...
for k, v := range ts.Timings() {
	fmt.Printf("%s: %d\n", k, v.Milliseconds())
}
```

### LINKED ISSUE #20968: Optimize e2e tests by parallelizing work in EnsureCleanState
# Summary

There can be steps to be executed in parallel (a naive assumption could be that nearly all steps can be) to save time.

# Motivation

EnsureCleanState is ~10-30% of each e2e test run time, so optimizing it can substantially reduce the overall e2e tests run time.

# Proposal

Figure out which blocks are safe to execute in parallel and do so.

### PR DESCRIPTION
**Pull request by andrii-korotkov-verkada**
Closes #20968
Closes #20967

A number of cleanup and create steps can be done in parallel, reducing the total runtime. Gather timing information as well and log it.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Comment by andrii-korotkov-verkada** [comment]
Seems to save a couple of minutes. The savings aren't as big as expected, since login as admin and config maps updates are taking some time, and the later doesn't parallelize well (sequential time is similar to parallel time).

Screenshot 2024-11-29 at 7 52 22 AM
Screenshot 2024-11-29 at 7 52 31 AM

**Comment by crenshaw-dev** [comment]
Can we remove the timing info code to simplify things now that you've done the initial speedup/analysis?

**Comment by andrii-korotkov-verkada** [comment]
> Can we remove the timing info code to simplify things now that you've done the initial speedup/analysis?

I'm not done yet :) I'm trying another commit to only do updates when necessary for slowest parts.
I'd rather keep them for me or other people who wanna optimize further, as it's not much added complexity. If you insist, I can remove though.

**Comment by crenshaw-dev** [comment]
I'd only leave it in place if there's also documentation on how to use the logging for further optimization _and_ if there's a strong likelihood that someone will actually use it to significantly improve performance. If the upper bound to improvement is ~30s, I say kill the code.

**Comment by andrii-korotkov-verkada** [comment]
Okay, I'll clean it up when done with optimizations here.

**Comment by andrii-korotkov-verkada** [comment]
Yay, best case time for login as admin was reduced to ~2ms instead of ~500-600ms due to skipping re-login if not necessary. Config map reset was reduced to ~80-200ms instead of ~800-1000ms due to skipping update if equivalent.
Screenshot 2024-11-29 at 11 36 40 AM

**Comment by andrii-korotkov-verkada** [comment]
That's another ~4 min of savings, bringing total savings to ~6 min.

Screenshot 2024-11-29 at 12 05 01 PM
Screenshot 2024-11-29 at 12 05 26 PM

**Comment by andrii-korotkov-verkada** [comment]
About the same times
Screenshot 2024-11-29 at 12 05 01 PM
Screenshot 2024-11-29 at 12 05 26 PM

**Comment by andrii-korotkov-verkada** [comment]
About the same times
Screenshot 2024-11-29 at 2 27 22 PM
Screenshot 2024-11-29 at 2 27 36 PM

**Comment by andrii-korotkov-verkada** [comment]
@crenshaw-dev, I think this is ready to go now.

File: test/e2e/fixture/fixture.go [hunk 1]
```diff
@@ -573,6 +593,21 @@ func WithTestData(testdata string) TestOption {
...
+func runFunctionsInParallel(t *testing.T, functions map[string]func()) {
+	t.Helper()
+
+	var wg sync.WaitGroup
+
+	for _, function := range functions {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			function()
```
Comment (crenshaw-dev) [inline test/e2e/fixture/fixture.go:605]: Are we relying on `os.Exit`s in the parallel functions for error handling?

Comment (andrii-korotkov-verkada) [reply test/e2e/fixture/fixture.go:605]: Hm, that's a good point, probably need to return the errors to the main function.

Comment (andrii-korotkov-verkada) [reply test/e2e/fixture/fixture.go:605]: It's not that easy. I've checked and it's relying on `os.Exit` after logging `Fatal`. Is that okay?

Comment (andrii-korotkov-verkada) [reply test/e2e/fixture/fixture.go]: I've figured out a way.

**Comment by andrii-korotkov-verkada** [comment]
With the latest update, it seems to be only a tiny bit faster, maybe winning like 20 seconds.

Before:
Screenshot 2024-12-01 at 4 36 34 PM
Screenshot 2024-12-01 at 4 36 39 PM

After:
Screenshot 2024-12-01 at 6 21 16 PM
Screenshot 2024-12-01 at 6 21 22 PM

File: test/e2e/fixture/applicationsets/utils/fixture.go [hunk 2]
```diff
@@ -113,33 +113,48 @@ func EnsureCleanState(t *testing.T) {
...
-	CheckError(fixtureClient.AppClientset.ArgoprojV1alpha1().Applications(TestNamespace()).DeleteCollection(context.Background(), v1.DeleteOptions{PropagationPolicy: &policy}, v1.ListOptions{}))
-
-	// kubectl delete secrets -l e2e.argoproj.io=true
-	CheckError(fixtureClient.KubeClientset.CoreV1().Secrets(TestNamespace()).DeleteCollection(context.Background(),
-		v1.DeleteOptions{PropagationPolicy: &policy}, v1.ListOptions{LabelSelector: TestingLabel + "=true"}))
+	fixture.RunFunctionsInParallelAndCheckErrors(t, map[string]func() error{
+		"delete_applications_resources_namespace": func() error {
+			// Delete the applicationset-e2e namespace, if it exists
+			err := fixtureClient.KubeClientset.CoreV1().Namespaces().Delete(context.Background(), ApplicationsResourcesNamespace, v1.DeleteOptions{PropagationPolicy: &policy})
+			if err != nil && !strings.Contains(err.Error(), "not found") { // 'not found' error is expected
```
Comment (crenshaw-dev) [inline test/e2e/fixture/applicationsets/utils/fixture.go:120]: I believe there's a k8s library function that checks whether an error is a "not found" error.

**Comment by crenshaw-dev** [comment]
So total savings here is around 5min?

**Comment by andrii-korotkov-verkada** [comment]
About 6-7 minutes.

**Comment by andrii-korotkov-verkada** [comment]
I've merged master in and let's wait for the test run.

**Comment by andrii-korotkov-verkada** [comment]
For some reason the tests for v1.29 and v1.28 were slower.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Last requested change :-)

File: util/errors/errors.go [hunk 3]
```diff
@@ -50,3 +51,7 @@ func Fatalf(exitcode int, format string, args ...interface{}) {
 	log.RegisterExitHandler(exitfunc)
 	log.Fatalf(format, args...)
 }
+
+func IsNotFound(err error) bool {
+	return apierr.IsNotFound(err)
+}
```
[the remark this answers is not in the document]
Comment (andrii-korotkov-verkada) [reply util/errors/errors.go]: Was a bit lazy to deal with imports for no good reason, just removed.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**